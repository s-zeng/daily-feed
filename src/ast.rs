use nonempty::NonEmpty;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ArticleContent {
    pub blocks: NonEmpty<ContentBlock>,
    pub reading_time_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct FeedContent {
    pub articles: NonEmpty<Article>,
    pub total_reading_time_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DocumentContent {
    pub feeds: NonEmpty<Feed>,
    pub total_reading_time_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Headline {
    pub title: String,
    pub published_date: Option<String>,
    pub source_name: String,
    pub url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Document {
    pub metadata: DocumentMetadata,
    pub front_page: Option<Vec<ContentBlock>>,
    pub content: Option<DocumentContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DocumentMetadata {
    pub title: String,
    pub author: String,
    pub description: Option<String>,
    pub generated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Feed {
    pub name: String,
    pub description: Option<String>,
    pub url: Option<String>,
    pub content: Option<FeedContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Article {
    pub title: String,
    pub metadata: ArticleMetadata,
    pub comments: Vec<Comment>,
    pub content: Option<ArticleContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ArticleMetadata {
    pub published_date: Option<String>,
    pub author: Option<String>,
    pub url: Option<String>,
    pub feed_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Comment {
    pub author: String,
    pub content: Vec<ContentBlock>,
    pub upvotes: u32,
    pub downvotes: u32,
    pub timestamp: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum ContentBlock {
    Paragraph(TextContent),
    Heading {
        level: u8,
        content: TextContent,
    },
    List {
        ordered: bool,
        items: Vec<TextContent>,
    },
    Quote(TextContent),
    Code {
        language: Option<String>,
        content: String,
    },
    Link {
        url: String,
        text: String,
    },
    Image {
        url: String,
        alt: Option<String>,
    },
    Raw(String), // For complex HTML that we want to preserve as-is
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TextContent {
    pub spans: Vec<TextSpan>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TextSpan {
    pub text: String,
    pub formatting: TextFormatting,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct TextFormatting {
    pub bold: bool,
    pub italic: bool,
    pub code: bool,
    pub link: Option<String>,
}

impl Document {
    pub fn new(title: String, author: String) -> Self {
        Self {
            metadata: DocumentMetadata {
                title,
                author,
                description: None,
                generated_at: chrono::Utc::now().to_rfc3339(),
            },
            front_page: None,
            content: None,
        }
    }

    pub fn set_content(&mut self, feeds: NonEmpty<Feed>, total_reading_time_minutes: u32) {
        self.content = Some(DocumentContent {
            feeds,
            total_reading_time_minutes,
        });
    }

    pub fn set_front_page(&mut self, front_page: Vec<ContentBlock>) {
        self.front_page = Some(front_page);
    }

    pub fn total_articles(&self) -> usize {
        self.content
            .as_ref()
            .map(|c| c.feeds.iter().map(|f| f.content.as_ref().map_or(0, |fc| fc.articles.len())).sum())
            .unwrap_or(0)
    }

    pub fn extract_headlines(&self) -> Vec<Headline> {
        self.content
            .as_ref()
            .map(|c| {
                c.feeds
                    .iter()
                    .flat_map(|feed| {
                        feed.content.as_ref().map(|fc| {
                            fc.articles.iter().map(|article| Headline {
                                title: article.title.clone(),
                                published_date: article.metadata.published_date.clone(),
                                source_name: article.metadata.feed_name.clone(),
                                url: article.metadata.url.clone(),
                            }).collect::<Vec<_>>()
                        }).unwrap_or_default()
                    })
                    .collect()
            })
            .unwrap_or_default()
    }
}

impl Feed {
    pub fn new(name: String) -> Self {
        Self {
            name,
            description: None,
            url: None,
            content: None,
        }
    }

    pub fn with_description(mut self, description: String) -> Self {
        self.description = Some(description);
        self
    }

    pub fn with_url(mut self, url: String) -> Self {
        self.url = Some(url);
        self
    }

    pub fn set_content(&mut self, articles: NonEmpty<Article>, total_reading_time_minutes: u32) {
        self.content = Some(FeedContent {
            articles,
            total_reading_time_minutes,
        });
    }
}

impl Article {
    pub fn new(title: String, feed_name: String) -> Self {
        Self {
            title,
            metadata: ArticleMetadata {
                published_date: None,
                author: None,
                url: None,
                feed_name,
            },
            comments: Vec::new(),
            content: None,
        }
    }

    pub fn set_content(&mut self, blocks: NonEmpty<ContentBlock>, reading_time_minutes: u32) {
        self.content = Some(ArticleContent {
            blocks,
            reading_time_minutes,
        });
    }

    pub fn with_published_date(mut self, date: String) -> Self {
        self.metadata.published_date = Some(date);
        self
    }

    pub fn with_url(mut self, url: String) -> Self {
        self.metadata.url = Some(url);
        self
    }

    pub fn add_comment(&mut self, comment: Comment) {
        self.comments.push(comment);
    }
}

impl TextContent {
    pub fn plain(text: String) -> Self {
        Self {
            spans: vec![TextSpan {
                text,
                formatting: TextFormatting::default(),
            }],
        }
    }

    pub fn from_spans(spans: Vec<TextSpan>) -> Self {
        Self { spans }
    }

    pub fn is_empty(&self) -> bool {
        self.spans.iter().all(|span| span.text.trim().is_empty())
    }

    pub fn to_plain_text(&self) -> String {
        self.spans.iter().map(|span| span.text.as_str()).collect()
    }
}

impl TextSpan {
    pub fn plain(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting::default(),
        }
    }

    pub fn bold(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                bold: true,
                ..Default::default()
            },
        }
    }

    pub fn italic(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                italic: true,
                ..Default::default()
            },
        }
    }

    pub fn code(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                code: true,
                ..Default::default()
            },
        }
    }

    pub fn link(text: String, url: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                link: Some(url),
                ..Default::default()
            },
        }
    }
}
