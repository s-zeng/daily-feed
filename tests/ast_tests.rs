use daily_feed::ast::*;
use nonempty::NonEmpty;

#[test]
fn test_document_creation() {
    let mut doc = Document::new("Test Document".to_string(), "Test Author".to_string());

    let mut feed = Feed::new("Test Feed".to_string()).with_description("A test feed".to_string());

    let mut article = Article::new("Test Article".to_string(), "Test Feed".to_string());
    article.set_content(
        NonEmpty::new(ContentBlock::Paragraph(TextContent::plain("Test content".to_string()))),
        0,
    );

    feed.set_content(NonEmpty::new(article), 0);
    doc.set_content(NonEmpty::new(feed), 0);

    // Filter out the timestamp for reproducible snapshots
    doc.metadata.generated_at = "2025-01-01T00:00:00.000000Z".to_string();

    insta::assert_json_snapshot!(doc);
}

#[test]
fn test_text_content_plain_text() {
    let content = TextContent::from_spans(vec![
        TextSpan::plain("Hello ".to_string()),
        TextSpan::bold("world".to_string()),
        TextSpan::plain("!".to_string()),
    ]);

    let plain_text = content.to_plain_text();
    insta::assert_snapshot!(plain_text);
}

#[test]
fn test_text_content_structure() {
    let content = TextContent::from_spans(vec![
        TextSpan::plain("Hello ".to_string()),
        TextSpan::bold("world".to_string()),
        TextSpan::plain("!".to_string()),
    ]);

    insta::assert_json_snapshot!(content);
}

#[test]
fn test_article_with_reading_time() {
    let article = Article {
        title: "Test Article".to_string(),
        metadata: ArticleMetadata {
            published_date: Some("2025-01-01".to_string()),
            author: Some("Test Author".to_string()),
            url: Some("https://example.com/article".to_string()),
            feed_name: "Test Feed".to_string(),
        },
        comments: vec![],
        content: Some(ArticleContent {
            blocks: NonEmpty::new(ContentBlock::Paragraph(TextContent::plain(
                "Test content".to_string(),
            ))),
            reading_time_minutes: 5,
        }),
    };

    insta::assert_json_snapshot!(article);
}

#[test]
fn test_article_without_reading_time() {
    let article = Article {
        title: "Test Article".to_string(),
        metadata: ArticleMetadata {
            published_date: Some("2025-01-01".to_string()),
            author: Some("Test Author".to_string()),
            url: Some("https://example.com/article".to_string()),
            feed_name: "Test Feed".to_string(),
        },
        comments: vec![],
        content: None,
    };

    insta::assert_json_snapshot!(article);
}

#[test]
fn test_feed_with_total_reading_time() {
    let feed = Feed {
        name: "Test Feed".to_string(),
        description: Some("Test description".to_string()),
        url: Some("https://example.com/feed".to_string()),
        content: Some(FeedContent {
            articles: NonEmpty::new(Article {
                title: "Article 1".to_string(),
                metadata: ArticleMetadata {
                    published_date: None,
                    author: None,
                    url: None,
                    feed_name: "Test Feed".to_string(),
                },
                comments: vec![],
                content: None,
            }),
            total_reading_time_minutes: 45,
        }),
    };

    insta::assert_json_snapshot!(feed);
}

#[test]
fn test_document_with_total_reading_time() {
    let doc = Document {
        metadata: DocumentMetadata {
            title: "Test Document".to_string(),
            author: "Test Author".to_string(),
            description: Some("Test description".to_string()),
            generated_at: "2025-01-01T00:00:00.000000Z".to_string(),
        },
        front_page: None,
        content: Some(DocumentContent {
            feeds: NonEmpty::new(Feed {
                name: "Test Feed".to_string(),
                description: None,
                url: None,
                content: Some(FeedContent {
                    articles: NonEmpty::new(Article {
                        title: "Test Article".to_string(),
                        metadata: ArticleMetadata {
                            published_date: None,
                            author: None,
                            url: None,
                            feed_name: "Test Feed".to_string(),
                        },
                        comments: vec![],
                        content: None,
                    }),
                    total_reading_time_minutes: 60,
                }),
            }),
            total_reading_time_minutes: 120,
        }),
    };

    insta::assert_json_snapshot!(doc);
}
