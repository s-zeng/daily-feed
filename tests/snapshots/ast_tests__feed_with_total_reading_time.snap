---
source: tests/ast_tests.rs
expression: feed
---
{
  "name": "Test Feed",
  "description": "Test description",
  "url": "https://example.com/feed",
  "content": {
    "articles": [
      {
        "title": "Article 1",
        "metadata": {
          "published_date": null,
          "author": null,
          "url": null,
          "feed_name": "Test Feed"
        },
        "comments": [],
        "content": null
      }
    ],
    "total_reading_time_minutes": 45
  }
}
